src/print.rs

```diff
@@ -1,66 +1,103 @@
+use colored::Colorize;
 use std::fmt;
 
 use crate::geo::location::Point;
 use crate::geo::orientation::Orientation;
+use crate::mission::Outcome;
 use crate::robot::{Command, Robot};
 
-pub fn robots<I>(stream: I)
+pub fn plan<I>(upper_right: Point, stream: I)
 where
-  I: Iterator<Item = (Robot, Vec<Command>)>,
+    I: Iterator<Item = (Robot, Vec<Command>)>,
 {
-  for (robot, commands) in stream {
-    println!("{}", robot);
-    println!(
-      "{}\n",
-      commands
-        .iter()
-        .map(|c| format!("{}", c))
-        .collect::<Vec<String>>()
-        .join("")
-    );
-  }
+    println!("{}", upper_right);
+    for (robot, commands) in stream {
+        println!("{}", robot);
+        println!(
+            "{}\n",
+            commands
+                .iter()
+                .map(|c| format!("{}", c))
+                .collect::<Vec<String>>()
+                .join("")
+        );
+    }
 }
```

Comment:
Do you think it'd be worth doing most of this (and the other print functions) as `impl std::fmt::Display for (Robot, Vec<Command>)` instead?  You could then just pull in itertools, which has [this format function](https://docs.rs/itertools/0.9.0/itertools/trait.Itertools.html#method.format) to handle the`Iterator` bit of it, and printing to stdout/stderr is fairly simple to add on top of that.

Since the `Display` you'd end up writing here is basically the reverse of your `MissionPlan` parser I think this would also let you test one in terms of the other:

    let test_case = generate();
    assert_eq!(test_case, test_case.to_string().parse());

(excuse the pseudocode, but hopefully you get the idea)

Reply:
I tried to implement `Display` for `Vec<Command>`, but wouldn't be allowed to, because it's a foreign trait for a foreign type.

I'll try your suggestions, thanks!

Reply:
Ah yeah, the orphan rule possibly still applies here tbh - never tried with tuples

Reply:
The workaround would presumably be to create a named type for `(Robot, Vec<Command>)` - though whether it's worthwhile is up to you

Reply:
I do use it often enough it might be worth it 🤔

Reply:
Confirmed, orphan rule still applies, but itertools were a super useful suggestion!